aip/client-libraries/4223.md

```diff
@@ -0,0 +1,39 @@
+---
+id: 4223
+state: draft # TODO(noahdietz): Update pre-merge
+created: 2022-05-11 # TODO(noahdietz): Update pre-merge
+---
+
+# Client-side paylod validation
+
+Services document the behavior of specific request fields using the
+`google.api.field_behavior` annotation defined in [AIP-203][]. This annotation
+conveys how a service interprets the field, including how a field may be
+validated (or not) by the service.
+
+## Guidance
+
+Client libraries **must not** implement client-side payload validation based
+on the `google.api.field_behavior` annotation, except to prevent
+[local failures](4223.md#local-failures).
+
+The `google.api.field_behavior` annotation is a primarily a machine-readable
+form of _documentation_, and **not** configuration for payload validation. The
+annotation conveys how the service _already_ handles the field, and alludes to
+a service's own payload validation or expectations.
+
+Admittedly, depending on services to implement payload validation means clients
+will make a network call that is bound to result in error. Furthermore, users
+then depend on services to provide actionable errors, when the client could have
+provided a language-idiomatic, localized error. However, robust server-side
+payload validation means that _all types_ of client (first party, third party,
+`curl`, etc.) will benefit from the **same** validation, and that there is no
+lag between service and clients when the field behavior changes.
+
+### Local failures
+
+Client libraris **may** implement client-side payload validation based on the
```

Comment:
I'm not sure if AIPs typically have examples, but it may be useful to show one here.

Reply:
I think I'd rather not provide an example as it is very language dependent (Language A might crash if a pointer is null but Language B might just default to an empty value). Is that fair?

Reply:
Yeah thats fine